crates/stages/stages/src/stages/headers.rs

```diff
@@ -155,11 +155,13 @@ where
 
         // If we only have the genesis block hash, then we are at first sync, and we can remove it,
         // add it to the collector and use tx.append on all hashes.
-        if let Some((hash, block_number)) = cursor_header_numbers.last()? {
-            if block_number.value()? == 0 {
-                self.hash_collector.insert(hash.key()?, 0)?;
-                cursor_header_numbers.delete_current()?;
-                first_sync = true;
+        if provider.tx_ref().entries::<RawTable<tables::HeaderNumbers>>()? == 1 {
+            if let Some((hash, block_number)) = cursor_header_numbers.last()? {
```

Comment:
because this table is keyed by hash, we definitely should check length as well

Reply:
I wonder if there's a perf diff first vs last but really doesn't matter here I assume